src/client.rs

```diff
@@ -77,6 +77,173 @@ pub extern "C" fn rustls_client_config_builder_build(
     }
 }
 
+/// Currently just a placeholder with no accessors yet.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.RootCertStore.html
+#[allow(non_camel_case_types)]
+pub struct rustls_root_cert_store {
+    _private: [u8; 0],
+}
+
+/// A representation of the rustls Certificate type, which is an array of
+/// bytes, nominally in DER-encoded X.509.
+/// https://docs.rs/rustls/0.19.0/rustls/struct.Certificate.html
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_certificate {
+    bytes: *const u8,
+    len: usize,
+}
+
+/// Input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+#[repr(C)]
+pub struct rustls_verify_server_cert_params {
+    end_entity: rustls_certificate,
+    intermediates: *const rustls_certificate,
+    intermediates_len: usize,
+    roots: *const rustls_root_cert_store,
+    dns_name: *const c_char,
+    dns_name_len: usize,
+    ocsp_response: *const u8,
+    ocsp_response_len: usize,
+}
+
+/// User-provided input to a custom certificate verifier callback. See
+/// rustls_client_config_builder_dangerous_set_certificate_verifier().
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_user_data = *mut libc::c_void;
+
+// According to the nomicon https://doc.rust-lang.org/nomicon/ffi.html#the-nullable-pointer-optimization):
+// > Option<extern "C" fn(c_int) -> c_int> is a correct way to represent a
+// > nullable function pointer using the C ABI (corresponding to the C type int (*)(int)).
+// So we use Option<...> here. This is the type that is passed from C code.
+#[allow(non_camel_case_types)]
+type rustls_verify_server_cert_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_verify_server_cert_user_data,
+        params: *const rustls_verify_server_cert_params,
+    ) -> rustls_result,
+>;
+
+// This is the same as a rustls_verify_server_cert_callback after unwrapping
+// the Option (which is equivalent to checking for null).
+type NonNullVerifyCallback = unsafe extern "C" fn(
+    userdata: rustls_verify_server_cert_user_data,
+    params: *const rustls_verify_server_cert_params,
+) -> rustls_result;
+
+// An implementation of rustls::ServerCertVerifier based on a C callback.
+struct Verifier {
+    callback: NonNullVerifyCallback,
+    userdata: rustls_verify_server_cert_user_data,
+}
+
+unsafe impl Send for Verifier {}
+unsafe impl Sync for Verifier {}
```

Comment:
What does it mean to provide an `unsafe impl`? Does this mean this type is _not_ actually `Send` or `Sync` but we're telling the compiler to pretend that it is?